### PR TITLE
Optimize discounted price calculation

### PR DESCRIPTION
**Pull request by kevin-eom**
What about pull out query for product.collection from loop?
I saw looped sql for product collection in my local log and it's getting many as much as discounts is set.

I want to merge this change because...

### Pull Request Checklist

1. [x] Privileged views and APIs are guarded by proper permission checks.
1. [x] All visible strings are translated with proper context.
1. [x] All data-formatting is locale-aware (dates, numbers, and so on).
1. [x] Database queries are optimized and the number of queries is constant.
1. [x] Database migration files are up to date.
1. [x] The changes are tested.
1. [x] GraphQL schema and type definitions are up to date.
1. [x] Changes are mentioned in the changelog.

### CODE REVIEW

File: saleor/discount/utils.py [hunk 1]
```diff
@@ -46,32 +46,26 @@ def remove_voucher_usage_by_customer(voucher, customer_email):
...
         return discount.sale.get_discount()
     raise NotApplicable(
         pgettext("Voucher not applicable", "Discount not applicable for this product")
     )
 
-
+    
 def get_product_discounts(product, discounts: Iterable[DiscountInfo]):
     """Return discount values for all discounts applicable to a product."""
+    product_collections = set(c.id for c in product.collections.all())
```
Comment (NyanKiyoshi) [inline saleor/discount/utils.py]:
```suggestion
    product_collections = product.collections.all().values_list("pk", flat=True)
```

File: saleor/discount/utils.py [hunk 2]
```diff
@@ -46,32 +46,26 @@ def remove_voucher_usage_by_customer(voucher, customer_email):
...
 
-def are_product_collections_on_sale(product, discount: DiscountInfo):
-    """Check if any collection is on sale."""
-    discounted_collections = discount.collection_ids
-    product_collections = set(c.id for c in product.collections.all())
-    return product_collections.intersection(discounted_collections)
-
-
-def get_product_discount_on_sale(product, discount: DiscountInfo):
+def get_product_discount_on_sale(product, product_collections, discount:DiscountInfo):
```
Comment (NyanKiyoshi) [inline saleor/discount/utils.py]: You need to update `test_sale_applies_to_correct_products` as well 👍

**Comment by maarcingebala** [comment]
@kevin-eom CI checks are failing for this PR: https://travis-ci.org/mirumee/saleor/builds/597590734

File: saleor/discount/utils.py (hunk 2 above)
Comment (kevin-eom) [reply saleor/discount/utils.py]: updated

**Comment by django-queries** [comment]
Here is the report for b151b91307bcfd5be8c9222720f3305b91c7fa5c (kevin-eom/saleor @ patch-1)
Base comparison is 1218f4ff14811c2145775df76cc06e18b37132b7.



<details><summary>No differences were found. (click me)</summary>
<p>

```diff
# api.benchmark checkout
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  add billing address to checkout            	         34	         34	             20
  add shipping to checkout                   	          7	          7	              0
  checkout payment charge                    	         14	         14	              0
  complete checkout                          	          6	          6	              0
  create checkout                            	         48	         48	             24

# api.benchmark homepage
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  retrieve main menu                         	          5	          5	              0
  retrieve product list                      	          4	          4	              0
  retrieve secondary menu                    	          5	          5	              0
  retrieve shop                              	          2	          2	              0

# api.benchmark product
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  product details                            	         15	         15	              3
  retrieve product attributes                	         13	         13	              2

# api.benchmark variant
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  retrieve variant list                      	         15	         15	              6

# api product sorting attributes
  test name                                  	left count 	right count	duplicate count
  -------------------------------------------	-----------	-----------	---------------
  sort product not having attribute data     	         21	         21	              0
```

</p>
</details>

**Comment by maarcingebala** [comment]
@kevin-eom This looks great. Could you apply `Black` and `flake8` code formatters to your changes? We use them to keep a consistent code style in the project.

**Comment by NyanKiyoshi** [comment]
@kevin-eom 
```
pip install black
black saleor/
```

Then in the future you should use `pre-commit` 👍 https://docs.saleor.io/docs/next/contributing/coding-style/

**Comment by maarcingebala** [comment]
@kevin-eom This PR looks good and I'd like to merge it, but I would only need you to apply correct `flake8` and `black` formatting. Will you have to fix that? Thanks!

**Comment by maarcingebala** [comment]
@kevin-eom Final call to become a Saleor contributor 😉. If no further activity occurs here, we'll open a PR with this code, since this is a useful change.

**Comment by maarcingebala** [comment]
Fixed in #4977

**Comment by juno05** [comment]
@maarcingebala Sorry for late response. I was away from work for a while.
And I'm glad to hear you implemented it. Thanks.

**Comment by kevin-eom** [comment]
@maarcingebala Sorry for late, I was away from work for while and couldn't help it out.
Thank you for your team worked for it and implemented it. :)

**Pull request closed without merging**